### PR TITLE
Create Meme and MemeDatabase objects

### PR DESCRIPTION
**Pull request by lattai**
Next Steps: Create admin webpage to create meme objects

### CODE REVIEW

File: src/main/java/com/google/sps/data/MemeDatabase.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  public Meme entityToMeme(Entity entity) {
+    String title = entity.getProperty(TITLE_QUERY_STRING).toString();
+    String source = entity.getProperty(SOURCE_QUERY_STRING).toString();
+    String id = entity.getProperty(ID_QUERY_STRING).toString();
+    Meme meme = new Meme(title, source, id);
+    return meme;
+  }
+
+  public List<Meme> getMemes() {
```
Comment (lzoebaker) [inline src/main/java/com/google/sps/data/MemeDatabase.java:54]: What is the use case for getting all the memes from the database?

File: src/test/java/com/google/sps/MemeDatabaseTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Rule public ExpectedException entityNotStored = ExpectedException.none();
+
+  @Test
+  public void addMemeToDatabase_successfulPath() {
```
Comment (lzoebaker) [inline src/test/java/com/google/sps/MemeDatabaseTest.java:52]: add a test for getting a specific meme by an id.

File: src/main/java/com/google/sps/data/MemeDatabase.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.*;
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/data/MemeDatabase.java:11]: Prefer to import the specific things you need and not .*

Importing lots of stuff can make your binary size increase which makes builds take longer and deployments be more fickle.

File: src/main/java/com/google/sps/data/MemeDatabase.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+// Database for all memes to be accessed
+public class MemeDatabase {
+  private DatastoreService datastore;
+  public List<Meme> memes = new ArrayList<>();
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/data/MemeDatabase.java:16]: Initialize in the constructor?

File: src/main/java/com/google/sps/data/MemeDatabase.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  public static final String SOURCE_QUERY_STRING = "source";
+  public static final String ID_QUERY_STRING = "id";
+
+  public MemeDatabase(DatastoreService datastore) {
+    this.datastore = datastore;
+    query = new Query(ENTITY_QUERY_STRING);
+  }
+
+  public void addMemeToDatabase(Meme meme) {
+    this.memes.add(meme);
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/data/MemeDatabase.java:29]: Why do we put this here? It looks like it's never accessed?

File: src/test/java/com/google/sps/MemeDatabaseTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  private static final String ID_2 = "234567";
+  private static final String ID_3 = "345678";
+  private static final String SOURCE_1 = "images/dir/source1.png";
+  private static final String SOURCE_2 = "images/source2.png";
+  private static final String SOURCE_3 = "images/source3.png";
+  private static final String TITLE_1 = "testTitle1";
+  private static final String TITLE_2 = "testTitle2";
+  private static final String TITLE_3 = "testTitle3";
+
+  @Before
```
Comment (WillHipschman) [inline src/test/java/com/google/sps/MemeDatabaseTest.java:40]: You should have one test per public method. Your class has 5 public methods so I would expect 5 tests here. I would then expect a 6th test exercising the the IllegalArgument case

File: src/test/java/com/google/sps/MemeDatabaseTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    Meme retrieved1 = this.memeDatabase.getMemeFromID(ID_1);
+
+    JsonElement result1 = JsonParser.parseString(gson.toJson(retrieved1));
+    JsonElement expected1 = JsonParser.parseString(gson.toJson(meme1));
+    Assert.assertEquals(result1, expected1);
+  }
+
+  @Test
+  public void queriedEntityNotStored_throwsEntityNotFoundException()
```
Comment (WillHipschman) [inline src/test/java/com/google/sps/MemeDatabaseTest.java:83]: Unit test should be named "method_condition_expectation". this should be "getEntityFromId_noEntity_throwsNotFound"

File: src/main/java/com/google/sps/data/MemeDatabase.java (hunk 3 above)
Comment (DirtGrubDylan) [reply src/main/java/com/google/sps/data/MemeDatabase.java:11]: +1

File: src/main/java/com/google/sps/data/MemeDatabase.java (hunk 4 above)
Comment (DirtGrubDylan) [reply src/main/java/com/google/sps/data/MemeDatabase.java:16]: Do we need `memes` at all?

File: src/test/java/com/google/sps/MemeDatabaseTest.java (hunk 6 above)
Comment (DirtGrubDylan) [reply src/test/java/com/google/sps/MemeDatabaseTest.java:40]: +1

File: src/main/java/com/google/sps/data/MemeDatabase.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    this.datastore = datastore;
+    query = new Query(ENTITY_QUERY_STRING);
+  }
+
+  public void addMemeToDatabase(Meme meme) {
+    try {
+      Entity memeEntity = memeToEntity(meme);
+      datastore.put(memeEntity);
+    } catch (IllegalArgumentException e) {
+      e.printStackTrace();
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/data/MemeDatabase.java:31]: Probably need to report something to the user? We're mostly ignoring this exception right now.

**Pull request merged**